src/basic/hostname-util.c

```diff
@@ -221,6 +221,39 @@ int sethostname_idempotent(const char *s) {
         return 1;
 }
 
+int shorten_overlong(const char *s, char **ret) {
+        char *h, *p;
+
+        /* Shorten an overlong name to HOST_NAME_MAX or to the first dot,
+         * whatever comes earlier. */
+
+        assert(s);
+
+        h = strdup(s);
+        if (!h)
+                return -ENOMEM;
+
+        if (strlen(h) <= HOST_NAME_MAX) {
+                *ret = h;
+                return 0;
+        }
+
+        for (p = h; *p; p++) {
+                if (*p == '.') {
+                        *p = 0;
+                        break;
+                }
+        }
+
+        strshorten(h, HOST_NAME_MAX);
+
+        if (strlen(h) == 0)
+                return -EFAULT;
```

Comment:
uh, EFAULT is a strange choice i think, it's has a very clear meaning regarding memory issues which might potentially happen here too. Pick any other error, for example -EDOM or so...

Reply:
also, `h` is leaked here

Reply:
`EFAULT` was to mimic `sethostname(2)`, but I just realized I completely misread its meaning. I'll go for -EDOM.